From d7158dcf0086572992fab02fbd50b286506f2de0 Mon Sep 17 00:00:00 2001
From: Gilles Dubreuil <gilles@redhat.com>
Date: Tue, 18 Nov 2014 17:01:40 +1100
Subject: [PATCH] Fixed issue with vlan on redhat provider

When interface is a vlan, the VLAN=yes stanza must stick to interface
configuration file and shouldn't be present on bridge ifcfg file either

RHBZ#1160418

Change-Id: If1c40927f7e1b26b2ae6a7eb33f6d76d41983fc3
---
 vswitch/lib/puppet/provider/vs_port/ovs_redhat.rb | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/vswitch/lib/puppet/provider/vs_port/ovs_redhat.rb b/vswitch/lib/puppet/provider/vs_port/ovs_redhat.rb
index f2cf88d..2bc4374 100644
--- a/vswitch/lib/puppet/provider/vs_port/ovs_redhat.rb
+++ b/vswitch/lib/puppet/provider/vs_port/ovs_redhat.rb
@@ -39,6 +39,9 @@ Puppet::Type.type(:vs_port).provide(:ovs_redhat, :parent => :ovs) do
       end
 
       port = IFCFG::Port.new(@resource[:interface], @resource[:bridge])
+      if vlan?
+        port.set('VLAN' => 'yes')
+      end
       port.save(BASE + @resource[:interface])
 
       bridge = IFCFG::Bridge.new(@resource[:bridge], template)
@@ -115,6 +118,17 @@ Puppet::Type.type(:vs_port).provide(:ovs_redhat, :parent => :ovs) do
         items.merge!(m[1] => m[2])
       end
     end
+    items.delete('VLAN') if items.has_key?('VLAN')
     items
   end
+
+  def vlan?
+    if File.read('/proc/net/vlan/config') =~ /#{@resource[:interface]}/
+      return true
+    else
+      return false
+    end
+  rescue Errno::ENOENT
+    return false
+  end
 end
